WP3/D3.8/report.tex

```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
-% leave it to the conclusion.
+The SageMath project had traditionally employed two distinct CI systems, the
+Buildbot and Sage's own development, the
+patchbot\footnote{\url{https://github.com/sagemath/sage-patchbot}}.
+
+\subsubsection{Buildbot and patchbot}
+Sage's Buildbot runs as a non-advisory CI system on a wide variety of
+platforms. It builds Sage and runs its test suite for every change that gets
+accepted into the mainline of the Sage code. Its main purpose is not to be an
+advisory tool for developers but to catch problems and incompatibilities of
+Sage or its dependencies, in particular on more obscure platforms that
+dependent projects do often not test for. Due to the vast range of platforms
+covered, the Buildbot can not realistically be turned into an advisory CI
+system as it is not conceivable to provision and administer the number of
+machines that would be necessary to run Sage's test suite at every stage of a
+test that gets proposed by a developer.
+
+Therefore, the Sage community at some point developed its own patchbot system
+which runs Sage's test suite at every stage of a proposed change. Anybody can
+run a patchbot on their private computer, typically a Linux system. These
+patchbots will then look for proposed changes to Sage, run Sage's test suite,
+and report findings back to a central server. It is worth noting that the
+patchbot system predates most of the currently popular CI systems.
+
+The patchbot system comes with a number of issues: It is not nearly as polished
+as the professional offerings which means that its workings can often be
+confusing and that it also requires continuous maintenance and development from
+the SageMath community and by those that run the patchbots on their private
+machines. The fact that it is running on private machines means that due to
+load, connectivity issues, and other temporary issues on these machines, the
+patchbots are very frequently reporting false positives. This noise can be very
+frustrating as it makes the patchbot system unreliable and in particular harder
+to understand for new contributors. Finally, due to limitations of the patchbot
+architecture, not all types of changes can be tested, in particular it can not
+test changes that involve dependent software packages. This can be frustrating
+as testing such changes is often the most time consuming task for developers
+and it would be particularly nice to give such tasks away to a CI system.
+
+\subsubsection{Modern CI with GitLab CI and CircleCI}
+To fix these issues we were looking at best pratices that would give the
+SageMath community fast, reproducible, reliable CI with minimal maintenance and
+continuous development necessary to keep the CI working for everybody.  We
+settled for GitLab CI, a modern open-source CI system, and CircleCI, a popular
+closed-source system. Which one to use is up to each developer. For simplicity
+we focus on GitLab CI here. GitLab CI typically uses docker containers as its
```

Comment:
Perhaps worth stating explicitly that most of what we write about GitLab here applies fairly similarly to Circle CI.  Also I would be interested in a note about why you chose to try two systems in the first place, and whether we intend to maintain both systems long term, or if it's more a question of evaluating them both and then focusing on whichever one works better.

Reply:
I can just fix this during final editing, but we should write "Docker" not "docker".

Reply:
GitLab CI did not integrate with GitHub back then. Now it does I heard, so maybe there is not that much of a point anymore. You can also SSH into the machines running jobs on CircleCI. GitLab is actively working on that feature but they don't have it yet. Then again that's not so terribly important anymore since I am now basing everything on the docker images so you can just download these and run everything yourself if you want to debug things.
So, we'll probably drop CircleCI support again in the long run.